### ISSUE TITLE
Helm rollback fail when resource (StatefulSet) no longer "monitored" by Helm

### ISSUE THREAD

**Issue by etedpet** [helm/helm#8741]
I'm having a stateful service (DB) that today is deployed using a StatefulSet. The service should be evolved so to include an _operator_ that operates the DB.
In the helm chart for the new service (new version) there is no template that contains a StatefulSet. Instead the operator is deployed together with an operand CR. The operator then reads the CR and creates/updates a StatefulSet - This all works fine.

But I want to have a _smooth migration_ from old chart to the new (operator-based) chart - so that there will be no downtime of the DB service during upgrade!
I've solved this with the help from 2 batch jobs (pre/post-upgrade hooks) to set/remove the annotation: `helm.sh/resource-policy=keep`
This way the StatefulSet will be kept post upgrade (even though no StatefulSet is part of the upgraded chart) and the operator will "take over" control of the StatefulSet. This works fine since the namespace and name is the same...
If the release is uninstalled then the CR will be deleted and the operator will then delete the StatefulSet.

This all works fine so now over to the problem - When I try to do a _helm rollback_ things crash:
```
> helm rollback app
Error: no StatefulSet with the name "test" found
```
But a "downgarde" (`helm upgrade` to the previous version) works fine!

To illustrate the problem I've constructed a simpler example to reproduce the issue (the manifests for the 2 charts are pasted below).

_test-chart1_ contains StatefulSet with the annotation "helm.sh/resource-policy": keep
_test-chart2_ contains only a Service (not the StatefulSet)

Steps to reproduce:

**1** - Install:
```
> helm install app test-chart1
> kubectl get all
NAME         READY   STATUS    RESTARTS   AGE
pod/test-0   1/1     Running   0          1m52s
pod/test-1   1/1     Running   0          1m50s
pod/test-2   1/1     Running   0          1m49s

NAME                    READY   AGE
statefulset.apps/test   3/3     1m53s
```

**2** - Upgrade: (note that the StatefulSet is kept)
```
> helm upgrade app test-chart2
> kubectl get all

NAME         READY   STATUS    RESTARTS   AGE
pod/test-0   1/1     Running   0          2m52s
pod/test-1   1/1     Running   0          2m50s
pod/test-2   1/1     Running   0          2m49s

NAME                   TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)   AGE
service/test-service   ClusterIP   10.96.45.192   <none>        80/TCP    2m26s

NAME                    READY   AGE
statefulset.apps/test   3/3     2m53s
```
**3** - Rollback:
```
> helm history app 
REVISION	UPDATED                 	STATUS    	CHART           	APP VERSION	DESCRIPTION                                                     
1       	Wed Sep 16 16:37:58 2020	superseded	test-chart-0.1.0	1.16.0     	Install complete                                                
2       	Wed Sep 16 16:38:23 2020	superseded	test-chart-0.2.0	1.18.0     	Upgrade complete 

> helm rollback app --debug 
rollback.go:60: [debug] preparing rollback of app
rollback.go:108: [debug] rolling back app (current: v2, target: v1)
rollback.go:67: [debug] creating rolled back release for app
rollback.go:73: [debug] performing rollback of app
client.go:173: [debug] checking 1 resources for changes
rollback.go:166: [debug] warning: Rollback "app" failed: no StatefulSet with the name "test" found
Error: no StatefulSet with the name "test" found
helm.go:84: [debug] no StatefulSet with the name "test" found
helm.sh/helm/v3/pkg/kube.(*Client).Update.func1
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/pkg/kube/client.go:201
helm.sh/helm/v3/pkg/kube.ResourceList.Visit
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/pkg/kube/resource.go:32
helm.sh/helm/v3/pkg/kube.(*Client).Update
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/pkg/kube/client.go:174
helm.sh/helm/v3/pkg/action.(*Rollback).performRollback
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/pkg/action/rollback.go:162
helm.sh/helm/v3/pkg/action.(*Rollback).Run
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/pkg/action/rollback.go:74
main.newRollbackCmd.func1
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/cmd/helm/rollback.go:59
github.com/spf13/cobra.(*Command).execute
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842
github.com/spf13/cobra.(*Command).ExecuteC
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main
	/home/jenkins/erikube/slave-root/workspace/gic-build-v4/components/helmv3/src/k8s.io/helm/cmd/helm/helm.go:83
runtime.main
	/usr/local/go/src/runtime/proc.go:203
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1357
```

The error message `Error: no StatefulSet with the name "test" found` is a bit confusing - because there really exist a StatefulSet with that name... But after looking into the code (client.go # Update(original, target ResourceList, force bool) it fails because a StatefulSet is found even though the current configuration (original) says it should not...

But I'm hoping that this is something that SHOULD work, right?

Manifests:
```
> helm get all app --revision 1
NAME: app
LAST DEPLOYED: Wed Sep 16 16:37:58 2020
NAMESPACE: mynamespace
STATUS: superseded
REVISION: 1
TEST SUITE: None
USER-SUPPLIED VALUES:

COMPUTED VALUES:
image: nginx:latest
nameOverride: test
replicaCount: 1

HOOKS:
MANIFEST:
---
# Source: test-chart/templates/test-db-statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: test
  annotations:
    "helm.sh/resource-policy": keep
spec:
  selector:
    matchLabels:
      app: test
  serviceName: test
  replicas: 3
  template:
    metadata:
      labels:
        app: test
    spec:
      terminationGracePeriodSeconds: 10
      containers:
      - name: mydb
        image: nginx:latest   # <--- Just a "test image"
        ports:
        - containerPort: 11211



> helm get all app --revision 2
NAME: app
LAST DEPLOYED: Wed Sep 16 16:38:23 2020
NAMESPACE: mynamespace
STATUS: superseded
REVISION: 2
TEST SUITE: None
USER-SUPPLIED VALUES:

COMPUTED VALUES:
image: nginx:latest
nameOverride: test
replicaCount: 3

HOOKS:
MANIFEST:
---
# Source: test-chart2/templates/test-service.yaml
apiVersion: v1
kind: Service
metadata:
  name: test-service
spec:
  selector:
    app: MyApp
  ports:
    - protocol: TCP
      port: 80
      targetPort: 9376
```

Output of `helm version` (same issue on master branch):
```
version.BuildInfo{Version:"v3.2.1", GitCommit:"8e8300b2a3c43829dd8a8b329186fcf96da9f4a6", GitTreeState:"clean", GoVersion:"go1.13.10"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:51:57Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"06ad960bfd03b39c8310aaf92d1e7c12ce618213", GitTreeState:"clean", BuildDate:"2020-02-16T17:47:13Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by bacongobbler** [comment]
I believe this is a duplicate of #8228. Have you tried using https://github.com/helm/helm/pull/8683 and see if that fixes things?

I was just looking at this bug yesterday. There appears to be an issue when trying to fetch a resource from Kubernetes when the annotation has changed.

**Comment by bacongobbler** [comment]
Yep. The use case and bug report are identical to #8228. Therefore, I'm going to close this as a duplicate of that issue. Thank you for the report.

**Comment by etedpet** [comment]
I tried the fix #8683, but it did not solve this issue. Still got the same error during rollback.

```
> helm-latest version
version.BuildInfo{Version:"v3.3+unreleased", GitCommit:"68d0483890f0a63f784f0f014e032b3a5f40e149", GitTreeState:"clean", GoVersion:"go1.15"}

> helm-latest rollback app --debug 
rollback.go:63: [debug] preparing rollback of app
rollback.go:111: [debug] rolling back app (current: v2, target: v1)
rollback.go:70: [debug] creating rolled back release for app
rollback.go:76: [debug] performing rollback of app
client.go:174: [debug] checking 1 resources for changes
rollback.go:169: [debug] warning: Rollback "app" failed: no StatefulSet with the name "test" found
Error: no StatefulSet with the name "test" found
helm.go:94: [debug] no StatefulSet with the name "test" found
helm.sh/helm/v3/pkg/kube.(*Client).Update.func1
	/local/repos/helm/pkg/kube/client.go:216
helm.sh/helm/v3/pkg/kube.ResourceList.Visit
	/local/repos/helm/pkg/kube/resource.go:32
helm.sh/helm/v3/pkg/kube.(*Client).Update
	/local/repos/helm/pkg/kube/client.go:175
helm.sh/helm/v3/pkg/action.(*Rollback).performRollback
	/local/repos/helm/pkg/action/rollback.go:165
helm.sh/helm/v3/pkg/action.(*Rollback).Run
	/local/repos/helm/pkg/action/rollback.go:77
main.newRollbackCmd.func2
	/local/repos/helm/cmd/helm/rollback.go:69
github.com/spf13/cobra.(*Command).execute
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842
github.com/spf13/cobra.(*Command).ExecuteC
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950
github.com/spf13/cobra.(*Command).Execute
	/home/etedpet/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main
	/local/repos/helm/cmd/helm/helm.go:93
runtime.main
	/usr/local/go/src/runtime/proc.go:204
runtime.goexit
	/usr/local/go/src/runtime/asm_amd64.s:1374
```
